File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Required_10/_95_Validate_Binary_Search_Tree_Medium.java
package J_3_Binary_Tree_Divide_Conquer.Required_10;
import java.util.*;
/** 95. Validate Binary Search Tree
 * Medium

 * Created by tianhuizhu on 6/27/17.
 */
public class _95_Validate_Binary_Search_Tree_Medium {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    // version 1 Traverse
    public class Solution1 {
        private int lastVal = Integer.MIN_VALUE;
        private boolean firstNode = true;
        public boolean isValidBST(TreeNode root) {
            if (root == null) {
                return true;
            }
            if (!isValidBST(root.left)) {
                return false;
            }
            if (!firstNode && lastVal >= root.val) {
                return false;
            }
            firstNode = false;
            lastVal = root.val;
            if (!isValidBST(root.right)) {
                return false;
            }
            return true;
        }
    }


    // version 2  Divide and Conquer
    class ResultType {
        boolean is_bst;
        int maxValue, minValue;

        ResultType(boolean is_bst, int maxValue, int minValue) {
            this.is_bst = is_bst;
            this.maxValue = maxValue;
            this.minValue = minValue;
        }
    }

    public class Solution2 {
        /**
         * @param root: The root of binary tree.
         * @return: True if the binary tree is BST, or false
         */
        public boolean isValidBST(TreeNode root) {
            ResultType r = validateHelper(root);
            return r.is_bst;
        }

        private ResultType validateHelper(TreeNode root) {
            if (root == null) {
                return new ResultType(true, Integer.MIN_VALUE, Integer.MAX_VALUE);
            }

            ResultType left = validateHelper(root.left);
            ResultType right = validateHelper(root.right);

            if (!left.is_bst || !right.is_bst) {
                // if is_bst is false then minValue and maxValue are useless
                return new ResultType(false, 0, 0);
            }

            if (root.left != null && left.maxValue >= root.val ||
                    root.right != null && right.minValue <= root.val) {
                return new ResultType(false, 0, 0);
            }

            return new ResultType(true,
                    Math.max(root.val, right.maxValue),
                    Math.min(root.val, left.minValue));
        }
    }

    // version 3  Divide and Conquer
    public class Solution3 {
        /**
         * @param root: The root of binary tree.
         * @return: True if the binary tree is BST, or false
         */
        public boolean isValidBST(TreeNode root) {
            // write your code here
            return divConq(root, Long.MIN_VALUE, Long.MAX_VALUE);
        }

        private boolean divConq(TreeNode root, long min, long max){
            if (root == null){
                return true;
            }
            if (root.val <= min || root.val >= max){
                return false;
            }
            return divConq(root.left, min, Math.min(max, root.val)) &&
                    divConq(root.right, Math.max(min, root.val), max);
        }
    }
}

File: JS9Ch/src/J_3_Binary_Tree_Divide_Conquer/Required_10/python_cpp/_95_Validate_Binary_Search_Tree_Medium.cpp
/** 95. Validate Binary Search Tree
 * Medium*/


// version 1: traverse
class Solution {
private:
    TreeNode *lastNode = NULL;
public:
    /**
     * @param root: The root of binary tree.
     * @return: True if the binary tree is BST, or false
     */
    bool isValidBST(TreeNode *root) {
        if (root == NULL) {
            return true;
        }
        if (!isValidBST(root->left)) {
            return false;
        }
        if (lastNode != NULL && lastNode->val >= root->val) {
            return false;
        }
        lastNode = root;
        return isValidBST(root->right);
    }
};

// traverse 2: divide & conquer
class ResultType {
public:
    bool isBST;
    TreeNode *maxNode, *minNode;
    ResultType() {
        this->isBST = true;
        this->maxNode = NULL;
        this->minNode = NULL;
    }
};

class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: True if the binary tree is BST, or false
     */
    bool isValidBST(TreeNode *root) {
        ResultType result = helper(root);
        return result.isBST;
    }

    ResultType helper(TreeNode *root) {
        ResultType result;
        if (root == NULL) {
            return result;
        }

        ResultType left = helper(root->left);
        ResultType right = helper(root->right);

        if (!left.isBST || !right.isBST) {
            result.isBST = false;
            return result;
        }

        if (left.maxNode != NULL && left.maxNode->val >= root->val) {
            result.isBST = false;
            return result;
        }

        if (right.minNode != NULL && right.minNode->val <= root->val) {
            result.isBST = false;
            return result;
        }

        result.isBST = true;
        result.minNode = left.minNode == NULL ? root : left.minNode;
        result.maxNode = right.maxNode == NULL ? root : right.maxNode;
        return result;
    }
};